cpp/dolfinx/mesh/GraphBuilder.cpp

```diff
@@ -130,138 +131,154 @@ compute_local_dual_graph_keyed(
+        num_processes, (it.first)[0] - global_min, global_range);
 
     // Pack map into vectors to send
     send_buffer[dest_proc].insert(send_buffer[dest_proc].end(),
                                   it.first.begin(), it.first.end());
 
     // Add offset to cell numbers sent off process
-    send_buffer[dest_proc].push_back(it.second + offset);
+    send_buffer[dest_proc].push_back(it.second + cell_offset);
   }
 
   // Send data
   const graph::AdjacencyList<std::int64_t> received_buffer
       = dolfinx::MPI::all_to_all(
           mpi_comm, graph::AdjacencyList<std::int64_t>(send_buffer));
 
-  // Clear send buffer
-  send_buffer = std::vector<std::vector<std::int64_t>>(num_processes);
+  const int tdim = mesh::cell_dim(cell_type);
+  const int num_vertices_per_facet
+      = mesh::num_cell_vertices(mesh::cell_entity_type(cell_type, tdim - 1));
+
+  assert(received_buffer.array().size() % (num_vertices_per_facet + 1) == 0);
+  const int num_facets
+      = received_buffer.array().size() / (num_vertices_per_facet + 1);
 
-  // Map to connect processes and cells, using facet as key
-  typedef boost::unordered_map<std::vector<std::int64_t>,
-                               std::pair<std::int64_t, std::int64_t>>
-      MatchMap;
-  MatchMap matchmap;
+  const Eigen::Map<const Eigen::Array<std::int64_t, Eigen::Dynamic,
+                                      Eigen::Dynamic, Eigen::RowMajor>>
+      received_buffer_array(received_buffer.array().data(), num_facets,
+                            num_vertices_per_facet + 1);
 
-  // Look for matches to send back to other processes
-  std::pair<std::vector<std::int64_t>, std::pair<std::int64_t, std::int64_t>>
-      key;
-  key.first.resize(num_vertices_per_facet);
+  // Set up vector of owning processes for each received facet
+  const Eigen::Array<std::int32_t, Eigen::Dynamic, 1>& received_buffer_offsets
+      = received_buffer.offsets();
+  std::vector<int> proc(num_facets);
   for (int p = 0; p < num_processes; ++p)
   {
-    // Unpack into map
-    auto data_p = received_buffer.links(p);
-    for (int i = 0; i < data_p.rows(); i += (num_vertices_per_facet + 1))
-    {
-      // Build map key
-      std::copy(data_p.data() + i, data_p.data() + i + num_vertices_per_facet,
-                key.first.begin());
-      key.second.first = p;
-      key.second.second = data_p[i + num_vertices_per_facet];
+    for (int j = received_buffer_offsets[p] / (num_vertices_per_facet + 1);
+         j < received_buffer_offsets[p + 1] / (num_vertices_per_facet + 1); ++j)
+      proc[j] = p;
+  }
+
+  // Get permutation that takes facets into sorted order
+  std::vector<int> perm(num_facets);
+  std::iota(perm.begin(), perm.end(), 0);
+  // Lambda with capture for sort comparison
+  const auto cmp
+      = [&received_buffer_array, num_vertices_per_facet](int a, int b) {
+          const auto facet_a = received_buffer_array.row(a);
+          const auto facet_b = received_buffer_array.row(b);
```

Comment:
Need to take care that this doesn't create copies of the rows behind the scenes. Might be better to compute the offset manually.

Reply:
I'm sure it doesn't, but I'll change to using explicit pointers to make it clearer.